docs/migrations/support-slashed-denoms.md

```diff
@@ -0,0 +1,101 @@
+# Migrating from not supporing base denoms with slashes to supporting base denoms with slashes
+
+This document is intended to highlight significant changes which may require more information than presented in the CHANGELOG.
+Any changes that must be done by a user of ibc-go should be documented here.
+
+There are four sections based on the four potential user groups of this document:
+- Chains
+- IBC Apps
+- Relayers
+- IBC Light Clients
+
+This document is necessary when chains are upgrading from a version that does not support base denoms with slashes (e.g. v3.0.0) to a version that does (e.g. v3.1.0). All versions of ibc-go smaller than v1.5.0 for the v1.x release line, v2.3.0 for the v2.x release line, and v3.1.0 for the v3.x release line do *NOT** support IBC token transfers of coins whose base denoms contain slashes. Therefore the in-place of genesis migration described in this document are required when upgrading.
+
+If a chain receives coins of a base denom with slashes before it upgrades to supporting it, the receive may pass however the trace information will be incorrect.
+
+E.g. If a base denom of `testcoin/testcoin/testcoin` is sent to a chain that does not support slashes in the base denom, the receive will be successful. However, the trace information stored on the receiving chain will be: `Trace: "transfer/{channel-id}/testcoin/testcoin", BaseDenom: "testcoin"`.
+
+This incorrect trace information must be corrected when the chain does upgrade to fully supporting denominations with slashes.
+
+To do so, chain binaries should include a migration script that will run when the chain upgrades from not supporting base denominations with slashes to supporting base denominations with slashes.
+
+## Chains
+
+### ICS20 - Transfer
+
+The transfer module will now support slashes in base denoms, so we must iterate over current traces to check if any of them are incorrectly formed and correct the trace information.
+
+### Upgrade Proposal
+
+```go
+// Here the upgrade name is the upgrade name set by the chain
+app.UpgradeKeeper.SetUpgradeHandler("supportSlashedDenomsUpgrade",
+    func(ctx sdk.Context, _ upgradetypes.Plan, fromVM module.VersionMap) (module.VersionMap, error) {
+        // list of traces that must replace the old traces in store
+        var newTraces []ibctransfertypes.DenomTrace
+        app.TransferKeeper.IterateDenomTraces(ctx,
+            func(dt ibctransfertypes.DenomTrace) bool {
+                // check if the new way of splitting FullDenom
+                // into Trace and BaseDenom passes validation and
+                // is the same as the current DenomTrace.
+                // If it isn't then store the new DenomTrace in the list of new traces.
+                newTrace := ibctransfertypes.ParseDenomTrace(dt.GetFullDenomPath())
+                if err := newTrace.Validate(); err == nil && !reflect.DeepEqual(newTrace, dt) {
```

Comment:
Quick question, with the additionof the `reflect.DeepEqual` check, doesn't this mean that non-malformed traces that were previously in the store will be essentially discard here? 
Do we need the `reflect.DeepEqual` check?

This logic seems to differ from the Genesis Migration below.

Reply:
Ah apologies, we're not setting the whole list and overriding, its looping below and setting individually